### PR TITLE
feat: component token support algorithm

### PR DESCRIPTION
**Pull request by MadCcc**
- feat: component theme
- feat: component token support algorith,

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
https://github.com/ant-design/ant-design/discussions/43638

### 💡 Background and solution
https://preview-43810-ant-design.surge.sh/components/config-provider-cn#components-config-provider-demo-theme

https://github.com/ant-design/ant-design/assets/27722486/123591bc-7cb7-4164-a96f-97506899c54b

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |       Component Token support `algorithm` to calculate derivative tokens same as global.    |
| 🇨🇳 Chinese |    组件 Token 支持配置 `algorithm`，和全局 Token 一样可用算法计算派生。       |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 5121402</samp>

This pull request enhances the theme customization feature of the `ConfigProvider` component. It allows users to specify different theme tokens and derivation logic for individual components using the `theme` prop. It also updates the related hooks, context, interface, and demo files to support the new feature. It requires a newer version of `@ant-design/cssinjs` dependency.

### <samp>🤖 Generated by Copilot at 5121402</samp>

*  Add and update imports of theme-related types and functions from `@ant-design/cssinjs` and `components/theme` in several files ([link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaL1), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaL11-R10), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-7c4eca125442f4e9f24a1897a9eca97ed13cce775c5cb9e7e29f60ee574ede7cL1-R8), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cL5-R53))
*  Update the `ThemeConfig` interface and the `theme` prop of the `ConfigProvider` component to support component-specific theme tokens and algorithms ([link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaL32-R39), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dR19-R22), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dL20-R67), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L507-R507), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R513-R530), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-370b4f9c1f0527f049eb25b9235949218c2bb02e7c22c734d6ee7f068a4476d5L17-R28))
*  Implement the `getComputedToken` function and the `useTheme` hook to apply the theme derivation logic to the global and component themes ([link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-85ff925f8aa07cbb863b96b5bffc960720ad00516700939d2aa40cde582d41f3L15-R15), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-85ff925f8aa07cbb863b96b5bffc960720ad00516700939d2aa40cde582d41f3L53-L54), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cR74), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cL5-R53))
*  Update the `DesignTokenContext.Provider` component to pass the component-specific theme tokens and themes to the context consumers ([link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R539-R540))
*  Update the `package.json` file to use the latest version of `@ant-design/cssinjs` dependency ([link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L112-R112))
*  Add a demo file `components/config-provider/demo/theme.tsx` to showcase the new theme customization features for different components ([link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dL1-R12), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dR28-R30), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dL20-R67), [link](https://github.com/ant-design/ant-design/pull/43810/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dL46-R102))

### CODE REVIEW

**Review by li-jia-nan — APPROVED** [review]

File: components/config-provider/index.tsx [hunk 1]
```diff
@@ -512,11 +513,29 @@ const ProviderChildren: React.FC<ProviderChildrenProps> = (props) => {
...
+    const { algorithm, token, components, ...rest } = mergedTheme || {};
     const themeObj =
       algorithm && (!Array.isArray(algorithm) || algorithm.length > 0)
         ? createTheme(algorithm)
-        : undefined;
+        : defaultTheme;
+
+    const parsedComponents: any = {};
+    Object.entries(components || {}).forEach(([componentName, componentToken]) => {
+      parsedComponents[componentName] = { ...componentToken };
```
Comment (zombieJ) [inline components/config-provider/index.tsx:524]: `parsedComponents[componentName]` 到处都是，这里提个变量出来

File: components/theme/useToken.ts [hunk 2]
```diff
@@ -2,12 +2,55 @@ import type { Theme } from '@ant-design/cssinjs';
...
+  const { override, ...components } = overrideToken;
+
+  // Merge with override
+  let mergedDerivativeToken = {
+    ...derivativeToken,
+    override,
+  };
+
+  // Format if needed
+  mergedDerivativeToken = formatToken(mergedDerivativeToken);
```
Comment (zombieJ) [inline components/theme/useToken.ts:29]: formatToken 下面也调用了一次，这个确定是要这么做的么？

Comment (MadCcc) [reply components/theme/useToken.ts:29]: 只有这里调用了。cssinjs 里会被这个 getComputedToken 完全覆盖逻辑

Comment (MadCcc) [reply components/theme/useToken.ts:29]: 所以理论上 useCacheToken 是可以不传 formatToken 的

**Pull request merged**